workers/agent/agent.go

```diff
@@ -0,0 +1,686 @@
+package agent
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/signal"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/gaia-pipeline/gaia"
+	"github.com/gaia-pipeline/gaia/helper/filehelper"
+	"github.com/gaia-pipeline/gaia/helper/pipelinehelper"
+	"github.com/gaia-pipeline/gaia/store"
+	"github.com/gaia-pipeline/gaia/workers/agent/api"
+	pb "github.com/gaia-pipeline/gaia/workers/proto"
+	"github.com/gaia-pipeline/gaia/workers/scheduler"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/metadata"
+)
+
+const (
+	// schedulerTickerSeconds defines the interval in seconds for the scheduler.
+	schedulerTickerSeconds = 3
+
+	// updateTickerSeconds defines the interval in seconds to send updates.
+	updateTickerSeconds = 2
+
+	// chunkSize is the size of binary chunks transferred to workers.
+	chunkSize = 64 * 1024 // 64 KiB
+
+	// idMDKey is the key used for the gRPC metadata map.
+	idMDKey = "uniqueid"
+
+	// defaultHostname is the default hostname set for the mTLS certificate.
+	defaultHostname = "gaia-pipeline.io"
+)
+
+// Agent represents an instance of an agent
+type Agent struct {
+	// client represents the interface for the worker client
+	client pb.WorkerClient
+
+	// self represents the current agent instance information
+	self *pb.WorkerInstance
+
+	// certFile represents the local path to the agent cert
+	certFile string
+
+	// keyFile represents the local path to the agent key
+	keyFile string
+
+	// caCertFile represents the local path to the agent ca cert
+	caCertFile string
+
+	// Instance of scheduler
+	scheduler scheduler.GaiaScheduler
+
+	// Instance of store
+	store store.GaiaStore
+
+	// Signal channel for this agent
+	sigs chan os.Signal
+}
+
+// InitAgent initiates the agent instance
+func InitAgent(scheduler scheduler.GaiaScheduler, store store.GaiaStore) *Agent {
+	ag := &Agent{
+		scheduler: scheduler,
+		store:     store,
+	}
+
+	// Set path to local certificates
+	ag.certFile = filepath.Join(gaia.Cfg.HomePath, "cert.pem")
```

Comment:
Shouldn't this be a separate configuration entity since agents can be in a different location then gaia Home?

Reply:
Agents use `HomePath` to store their local data which must be different from the primary home directory. I just thought there is no reason to create a new path variable for the agents.